### ISSUE TITLE
GHC 9.12

### ISSUE THREAD

**Issue by andreasabel** [haskell/aeson#1124]
For GHC 9.12, aeson needs bumps of `ghc-prim` and `template-haskell`.
> [__5] rejecting: time-1.14/installed-230c (conflict: aeson => template-haskell>=2.14.0.0 && <2.23, time => template-haskell==2.23.0.0/installed-cb92)
> [__7] rejecting: text-2.1.2/installed-0b0c (conflict: aeson => ghc-prim>=0.5.0.0 && <0.12, text => ghc-prim==0.13.0/installed-fd56)

**Comment by phadej** [comment]
Currently my work on updating to support GHC-9.12 is blocked on https://github.com/haskell/cabal/issues/10704. I would like to see that resolved in a way or another, to know whether I need to alter stuff or whether `cabal-install` will be fixed.

**Comment by phadej** [comment]
I can confirm the cabal issue is causing problems with `aeson` test suite: https://github.com/haskell/aeson/pull/1126

**Comment by erikd** [comment]
@phadej Would it be possible to do a metadata edit on Hackage to get the version there building with `ghc-9.12` ?

As in understand it only upper version bound changes are needed. Eg:
```
|diff --git a/aeson.cabal b/aeson.cabal
index 1b25c57..0c8666e 100644
--- a/aeson.cabal
+++ b/aeson.cabal
@@ -98,10 +98,10 @@ library
     , containers        >=0.6.0.1  && <0.8
     , deepseq           >=1.4.4.0  && <1.6
     , exceptions        >=0.10.4   && <0.11
-    , ghc-prim          >=0.5.0.0  && <0.12
-    , template-haskell  >=2.14.0.0 && <2.23
+    , ghc-prim          >=0.5.0.0  && <0.14
+    , template-haskell  >=2.14.0.0 && <2.24
     , text              >=1.2.3.0  && <1.3  || >=2.0 && <2.2
-    , time              >=1.8.0.2  && <1.13
+    , time              >=1.8.0.2  && <1.15
 
   -- Compat
   build-depends:
diff --git a/attoparsec-iso8601/attoparsec-iso8601.cabal b/attoparsec-iso8601/attoparsec-iso8601.cabal
index e824b09..f61c11e 100644
--- a/attoparsec-iso8601/attoparsec-iso8601.cabal
+++ b/attoparsec-iso8601/attoparsec-iso8601.cabal
@@ -44,7 +44,7 @@ library
     , base                >=4.12.0.0 && <5
     , integer-conversion  >=0.1      && <0.2
     , text                >=1.2.3.0  && <1.3.0.0 || >=2.0 && <2.2
-    , time                >=1.8.0.2  && <1.13
+    , time                >=1.8.0.2  && <1.15
     , time-compat         >=1.9.4    && <1.10
 
 source-repository head
diff --git a/cabal.project b/cabal.project
index 3210d21..40c1426 100644
--- a/cabal.project
+++ b/cabal.project
@@ -8,7 +8,6 @@ packages: benchmarks
 tests: true
 benchmarks: true
 
-allow-newer: hermes-json:attoparsec-iso8601
 
 -- packages: https://hackage.haskell.org/package/libperf-0.1/candidate/libperf-0.1.tar.gz
 -- packages: https://hackage.haskell.org/package/tasty-perfbench-0.1/candidate/tasty-perfbench-0.1.tar.gz
diff --git a/text-iso8601/text-iso8601.cabal b/text-iso8601/text-iso8601.cabal
index 0cfa523..2c6f3e5 100644
--- a/text-iso8601/text-iso8601.cabal
+++ b/text-iso8601/text-iso8601.cabal
@@ -47,7 +47,7 @@ library
       base                >=4.12.0.0 && <5
     , integer-conversion  >=0.1      && <0.2
     , text                >=1.2.3.0  && <1.3.0.0 || >=2.0 && <2.2
-    , time                >=1.8.0.2  && <1.13
+    , time                >=1.8.0.2  && <1.15
     , time-compat         >=1.9.4    && <1.10
 
 test-suite text-iso8601-tests
@@ -68,7 +68,7 @@ test-suite text-iso8601-tests
     , quickcheck-instances  >=0.3.29.1 && <0.4
     , tasty                 >=1.4.3    && <1.6
     , tasty-hunit           >=0.10.0.3 && <0.11
-    , tasty-quickcheck      >=0.10.2   && <0.11
+    , tasty-quickcheck      >=0.10.2   && <0.12
 
 benchmark text-iso8601-bench
   default-language: Haskell2010
@@ -86,4 +86,4 @@ benchmark text-iso8601-bench
   build-depends:
       attoparsec          >=0.14.4  && <0.15
     , attoparsec-iso8601  >=1.1.0.1 && <1.2
-    , tasty-bench         >=0.3.4   && <0.4
+    , tasty-bench         >=0.3.4   && <0.5
```

**Comment by phadej** [comment]
@erikd as you can see there is PR https://github.com/haskell/aeson/pull/1126 so writing something which is already done is not helpful.

As I said, I'm blocked on `cabal-install-3.14` having regressions. I want this regressions to be at least acknowledged (whether they will be fixed and when), to know how to proceed.

As far as I'm concerned, without `cabal-install-3.14` working, there is no build tool supporting GHC-9.12. So if you have energy and time, please help `cabal-install` maintainers.

**Comment by erikd** [comment]
As far as I can see, that regression in `cabal-install-3.14` does not affect people using `3.12` who would benefit from having the upper bounds fixed on Hackage.

**Comment by andreasabel** [comment]
I tend to agree with @erikd.
@phadej, you seem to want to use `aeson` as a lever to get things moving for `cabal-install`, but I am not sure this is helping the Haskell ecosystem overall.
Also, looking at
- https://github.com/haskell/cabal/issues/10704

there is already acknowledgement of this issue.  
Judging from the structure of the cabal release process, I don't think a hotfix will appear within the next days or weeks.

**Comment by phadej** [comment]
https://github.com/haskell-CI/haskell-ci/issues/766

**Comment by phadej** [comment]
And for the record

```
Warning: Unknown/unsupported 'ghc' version detected (Cabal 3.12.1.0 supports
'ghc' version < 9.12): /home/phadej/.ghcup/bin/ghc-9.12 is version 9.12.1
```

I really don't like this warning. If it's not an issue, I'm sure `Cabal` team could also release `3.12.2.0` with enough GHC-9.12 support to claim it's supported.

**Comment by taimoorzaeem** [comment]
[https://discourse.haskell.org/t/psa-correctness-issue-in-ghc-9-12/11204](https://discourse.haskell.org/t/psa-correctness-issue-in-ghc-9-12/11204)

> We plan to release a bug fix release 9.12.2 in the near future to address this issue and recommend people do not upgrade until then.

Maybe wait until `9.12.2` before next aeson release?

**Comment by coot** [comment]
I am getting a similar issue with `ghc-9.10.2` (but not `9.10.1`):
```
[__7] rejecting: aeson; 2.2.3.0, 2.2.2.0 (conflict: base => ghc-prim==0.12.0/installed-7419, aeson => ghc-prim>=0.5.0.0 && <0.12)
```
Both with `cabal-3.14.2.0` (locally) and `cabal-3.12.1.0` in [CI](https://github.com/input-output-hk/io-sim/actions/runs/15050356996/job/42303404053).

**Comment by phadej** [comment]
@coot have you `cabal update`d?

**Comment by coot** [comment]
Yes; the cI job runs `cabal update` just before building.

**Comment by phadej** [comment]
@coot in https://hackage.haskell.org/package/aeson-2.2.3.0/aeson.cabal there is 

```
ghc-prim          >=0.5.0.0  && <0.14
```

maybe you ahve `index-state` or somehow otherwise broken setup, but I'm 99% sure, the problem is on your end. I won't debug it for you
